zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -380,1147 +220,479 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
-        )?;
-        region.assign_advice(
-            || "tx_id_diff_inv",
-            self.tx_id_diff_inv,
-            calldata_offset,
-            || Value::known(tx_id_diff_inv),
-        )?;
-        region.assign_advice(
-            || "is_final",
-            self.is_final,
-            calldata_offset,
-            || Value::known(is_final),
-        )?;
-        region.assign_advice(
-            || "gas_cost",
-            self.calldata_gas_cost,
-            calldata_offset,
-            || Value::known(gas_cost),
-        )?;
-
-        let value_offset = 3 * tx_table_len;
-
-        region.assign_advice(
-            || "raw_pi.tx_value",
-            self.rpi,
-            offset + value_offset + self.circuit_block_len(),
-            || Value::known(tx_value),
-        )?;
+        let randomness = if keccak_hi_lo {
+            challenges.evm_word()
+        } else {
+            challenges.keccak_input()
+        };
+        let rpi = field_bytes
+            .iter()
+            .fold(Value::known(F::zero()), |acc, byte| {
+                acc.zip(t)
+                    .and_then(|(acc, t)| Value::known(acc * t + F::from(*byte as u64)))
+            });
+        let mut cells = vec![None; field_bytes.len() + 2];
+        for (i, byte) in field_bytes.iter().enumerate() {
+            let row_offset = *offset + i;
+
+            region.assign_fixed(
+                || "is_field_rlc",
+                self.is_field_rlc,
+                row_offset,
+                || Value::known(use_rlc),
+            )?;
 
-        // Add copy to vec
-        rpi_vals[offset + value_offset] = Value::known(tx_value);
+            // assign field bytes
+            let field_byte_cell = region.assign_advice(
+                || "field bytes",
+                self.rpi_field_bytes,
+                row_offset,
+                || Value::known(F::from(*byte as u64)),
+            )?;
+            field_rlc_acc = field_rlc_acc
+                .zip(t)
+                .and_then(|(acc, t)| Value::known(acc * t + F::from(*byte as u64)));
+            region.assign_advice(
+                || "field bytes acc",
+                self.rpi_field_bytes_acc,
+                row_offset,
+                || field_rlc_acc,
+            )?;
+            let rpi_cell = region.assign_advice(|| "rpi", self.rpi, row_offset, || rpi)?;
+            *rpi_rlc_acc = rpi_rlc_acc.zip(randomness).and_then(|(acc, randomness)| {
+                Value::known(acc * randomness + F::from(*byte as u64))
+            });
+            let rpi_rlc_acc_cell = region.assign_advice(
+                || "rpi_rlc_acc",
+                self.rpi_rlc_acc,
+                row_offset,
+                || *rpi_rlc_acc,
+            )?;
+            *rpi_len_acc += 1;
+            region.assign_advice(
+                || "rpi_len_acc",
+                self.rpi_len_acc,
+                row_offset,
+                || Value::known(F::from(*rpi_len_acc)),
+            )?;
```

Comment:
I don't find any constraints on `rpi_len_acc`, but I guess not necessary anymore anyway because you can just hardcode the length in the keccak lookup, so the column can be removed?

Reply:
Yeah, your are rigth, our public input has fixed length